### ISSUE TITLE
SSIM throws  error on γ values

### ISSUE THREAD

**Issue by mgautam98** [JuliaImages/ImageQualityIndexes.jl#16]
The SSIM throws error on small γ values
```
img = testimage("lena_gray_256") .|> Float64
noisy_img = img .+ 0.1 .* randn(size(img))

iqi = SSIM(KernelFactors.gaussian(1.5, 11), (0.0, 0.0448, 0.0448))  
assess(iqi, img, noisy_img)

ERROR: DomainError with -0.1274347376387049:
Exponentiation yielding a complex result requires a complex argument.
Replace x^y with (x+0im)^y, Complex(x)^y, or similar.
```

Problem is that σxy sometimes is negative and so `s` is -ve. And it throws DomainError, when taking power of -ve number.
https://github.com/JuliaImages/ImageQualityIndexes.jl/blob/c08443503e28f2bdb32620e9a4b95b71367abc0f/src/ssim.jl#L112-L114

**Comment by jiviteshjain** [comment]
According to the handbook [Structural Approaches to Image Quality Assessment](https://ece.uwaterloo.ca/~z70wang/publications/HIVP_chapter83.pdf),
> Notice that s(x, y) can take on negative values. As will be shown in later examples,
the negative structural similarity values correspond to the cases that the local
image structures are inverted.

However they do not talk about how to handle that with cases where gamma is less than 1.

**Comment by mgautam98** [comment]
A workaround for this maybe we can create a mask of signs and remove signs from `s` during `s^γ` and then multiply sign mask. 
@timholy @johnnychen94

**Comment by jiviteshjain** [comment]
According to [this](https://books.google.co.in/books?id=6Lc7DwAAQBAJ&pg=PA16&lpg=PA16&dq=structural+similarity+index+negative+covariance&source=bl&ots=GVXRvsdjZB&sig=ACfU3U3JzitSnpWseCektgVoAYeBDt9NUw&hl=en&sa=X&ved=2ahUKEwjWpcKxk4nqAhXvzTgGHXFEBcgQ6AEwEnoECAoQAQ#v=onepage&q=structural%20similarity%20index%20negative%20covariance&f=falsel) book, 
> Since the structure term can be negative, γ should be normalized to 1 to avoid any complex number and not to favor any anti-correlation. Alternatively, s(x, y) may be transformed to max(0, s(x, y)) to avoid any negative values.

I believe just saturating those values out at 0 makes more sense than artificially raising their magnitudes to that power.

**Comment by mgautam98** [comment]
By saturating at 0, how close do you mean because we can't saturate at exactly 0 because then whole ans will be 0 according to this eqn `l^α * c^β * s^γ`

**Comment by jiviteshjain** [comment]
> By saturating at 0, how close do you mean because we can't saturate at exactly 0 because then whole ans will be 0 according to this eqn `l^α * c^β * s^γ`

It will only be 0 for that particular element in the matrix. The final answer is the mean of all values in that matrix, so the effect should be small.

**Comment by johnnychen94** [comment]
> By saturating at 0, how close do you mean because we can't saturate at exactly 0 because then whole ans will be 0 according to this eqn `l^α * c^β * s^γ`

I think this is acceptable since we will take the mean of the ssim map; a pixel with 0 doesn't affect the overall result significantly.

I'm more interested in if other implementations(e.g., scikit-image, pyssim) has similar issues, if they already have a solution to this, then we'd better follow the same way to get equal ssim results across different implementations. If other implementations don't support this at all, then I think this is undefined and we could do whatever makes sense here.

**Comment by mgautam98** [comment]
Tensoflow is using RELU just like @jiviteshjain suggested for their [ssim_multiscale implementation](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/ops/image_ops_impl.py#L3766), 

pyssim is using [absolute everywhere](https://github.com/jterrace/pyssim/blob/ff9bd90c3eb7525013ad46babf66b7cc78391e89/ssim/ssimlib.py#L175) Just like I suggested
```
        # Compute the first term
        c1c2 = np.multiply(abs(cwtmatr1), abs(cwtmatr2))
        c1_2 = np.square(abs(cwtmatr1))
        c2_2 = np.square(abs(cwtmatr2))
        num_ssim_1 = 2 * np.sum(c1c2, axis=0) + self.k
        den_ssim_1 = np.sum(c1_2, axis=0) + np.sum(c2_2, axis=0) + self.k

        # Compute the second term
        c1c2_conj = np.multiply(cwtmatr1, np.conjugate(cwtmatr2))
        num_ssim_2 = 2 * np.abs(np.sum(c1c2_conj, axis=0)) + self.k
        den_ssim_2 = 2 * np.sum(np.abs(c1c2_conj), axis=0) + self.k
```
sklearn-image and pyssim only did it for α = β = γ = 1 so they didn't handled it.

**Comment by jiviteshjain** [comment]
Yes, that's true. I looked at the [tensorflow](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/ops/image_ops_impl.py#L3411), [scikit-image](https://github.com/scikit-image/scikit-image/blob/master/skimage/metrics/_structural_similarity.py), [pyssim](https://github.com/jterrace/pyssim/blob/master/ssim/ssimlib.py) as well as the paper's author's own Matlab [implementation](https://www.cns.nyu.edu/~lcv/ssim/ssim_index.m) of SSIM and none of them allows for custom values of alpha, beta or gamma. Because the default value of gamma = 1 is used, the problem of raising negative numbers to a power less than 1 does not arise.

The [tensorflow documentation](https://www.tensorflow.org/api_docs/python/tf/image/ssim) says that the output SSIM value can be in the range `(-1:1]`.

And the scikit-image implementation does allow for negative values in the `ssim_map` as well. Not a full-blown test case but I found negative values in their `ssim_map` matrix on a quick run on two images from their standard test image set:

![Screenshot from 2020-06-17 23-24-54](https://user-images.githubusercontent.com/43912285/84932674-471eda80-b0f2-11ea-8d0a-5628655c5866.png)

**Comment by johnnychen94** [comment]
Very nice work @mgautam98 and @jiviteshjain !

Looks like we don't have a standard here so I'll take #17 as the workaround to this (simply because this PR is proposed first). I believe users are not insensitive to these two implementations and it's also an "edge case" -- Indeed, most users should just use the default parameters (I just added a comment to the docstring in #18 to clarify this.)
